### PR TITLE
Guaranty IMAP session consistency

### PR DESCRIPTION
**Pull request by hoangdat**

### CODE REVIEW

**Comment by hoangdat** [comment]
```
[b27b6ca00f399e7410f3aad05909b526d8610e51] [ERROR] Failures: 
[b27b6ca00f399e7410f3aad05909b526d8610e51] [ERROR]   VacationIntegrationTest.jmapVacationShouldNotSendNotificationTwice:243->assertOneMessageReceived:393 1 expectation failed.
[b27b6ca00f399e7410f3aad05909b526d8610e51] JSON path [1][1].list doesn't match.
[b27b6ca00f399e7410f3aad05909b526d8610e51] Expected: a collection with size <1>
[b27b6ca00f399e7410f3aad05909b526d8610e51]   Actual: [{mailboxIds=[5], from={name=benwa@domain.tld, email=benwa@domain.tld}, textBody=Message explaining my wonderful vacations, id=5, to=[{name=matthieu@domain.tld, email=matthieu@domain.tld}]}, {mailboxIds=[5], from={name=benwa@domain.tld, email=benwa@domain.tld}, textBody=Message explaining my wonderful vacations, id=6, to=[{name=matthieu@domain.tld, email=matthieu@domain.tld}]}]
```
test this please

File: mailbox/event/event-rabbitmq/src/main/java/org/apache/james/mailbox/events/EventDispatcher.java [hunk 1]
```diff
@@ -80,4 +105,11 @@ void start() {
         return eventSerializer.toJson(event).getBytes(StandardCharsets.UTF_8);
     }
 
+    private OutboundMessage generateMessage(RoutingKeyConverter.RoutingKey routingKey, byte[] payload) {
+        Map<String, Object> headers = ImmutableMap.of(EVENT_BUS_ID, eventBusId.getId().toString());
```
Comment (chibenwa) [inline mailbox/event/event-rabbitmq/src/main/java/org/apache/james/mailbox/events/EventDispatcher.java]: Can we have a `EventBusId::asString` for doing this and not leaking the UID representatiopn?

File: mailbox/event/event-rabbitmq/src/main/java/org/apache/james/mailbox/events/EventDispatcher.java [hunk 2]
```diff
@@ -80,4 +105,11 @@ void start() {
...
+        AMQP.BasicProperties.Builder propertiesBuilder = new AMQP.BasicProperties.Builder();
```
Comment (chibenwa) [inline mailbox/event/event-rabbitmq/src/main/java/org/apache/james/mailbox/events/EventDispatcher.java]: Can this be of type `AMQP.BasicProperties` sot that we can only:

```
return new OutboundMessage(MAILBOX_EVENT_EXCHANGE_NAME, routingKey.asString(), properties, payload);
```

File: mailbox/event/event-rabbitmq/src/main/java/org/apache/james/mailbox/events/KeyRegistrationHandler.java [hunk 3]
```diff
@@ -106,18 +108,29 @@ Registration register(MailboxListener listener, RegistrationKey key) {
...
                 .doOnError(e -> LOGGER.error("Exception happens when handling event of user {}", event.getUser().asString(), e))
-                .onErrorResume(e -> Mono.empty()))
-            .subscribeOn(Schedulers.elastic())
+                .onErrorResume(e -> Mono.empty())
+                .then().cache())
+            .subscribeOn(Schedulers.parallel())
             .then();
     }
 
+    private boolean checkInvalidListener(EventBusId eventBusId, MailboxListener listener) {
```
Comment (chibenwa) [inline mailbox/event/event-rabbitmq/src/main/java/org/apache/james/mailbox/events/KeyRegistrationHandler.java]: I propose `skipLocalSynchronousListeners` as a method name.

File: mailbox/event/event-rabbitmq/src/main/java/org/apache/james/mailbox/events/KeyRegistrationHandler.java [hunk 4]
```diff
@@ -117,7 +117,7 @@ Registration register(MailboxListener listener, RegistrationKey key) {
         Event event = toEvent(delivery);
 
         return mailboxListenerRegistry.getLocalMailboxListeners(registrationKey)
-            .filter(listener -> !checkInvalidListener(eventBusId, listener))
+            .filter(listener -> !skipLocalSynchronousListeners(eventBusId, listener))
```
Comment (chibenwa) [inline mailbox/event/event-rabbitmq/src/main/java/org/apache/james/mailbox/events/KeyRegistrationHandler.java]: Or `-> ! isLocalSynchronousListener(...)`

**Comment by hoangdat** [comment]
```
[642cb6d01dc430fc00f108746cf99c3475decc48] [ERROR] Failures: 
[642cb6d01dc430fc00f108746cf99c3475decc48] [ERROR]   SessionWithInitializedTablesFactoryTest.createSessionShouldKeepTheSetSchemaVersionWhenTypesAndTablesHaveNotChanged:96 expected:<...hemaVersion{version=[6]}> but was:<...hemaVersion{version=[2]}>
[642cb6d01dc430fc00f108746cf99c3475decc48] [INFO] 
[642cb6d01dc430fc00f108746cf99c3475decc48] [ERROR] Tests run: 116, Failures: 1, Errors: 0, Skipped: 0
```
test this please

File: mailbox/api/src/test/java/org/apache/james/mailbox/events/KeyContract.java [hunk 5]
```diff
@@ -346,5 +346,17 @@ default void localDispatchedListenersShouldBeDispatchedWithoutDelay() throws Exc
             verify(mailboxListener1, timeout(ONE_SECOND).times(1)).event(any());
         }
 
+        @Test
+        default void registerShouldNotDispatchPastEventsFromOtherEventBus() throws Exception {
```
[the remark this answers is not in the document]
Comment (hoangdat) [reply mailbox/api/src/test/java/org/apache/james/mailbox/events/KeyContract.java]: removed

**Comment by trantienduchn** [comment]
rebased on master, resloved confict

File: mailbox/event/event-rabbitmq/src/main/java/org/apache/james/mailbox/events/EventDispatcher.java [hunk 6]
```diff
@@ -54,12 +68,23 @@ void start() {
...
 
     Mono<Void> dispatch(Event event, Set<RegistrationKey> keys) {
+        Mono<Void> localListenerDelivery = Flux.fromIterable(keys)
+            .subscribeOn(Schedulers.elastic())
+            .flatMap(mailboxListenerRegistry::getLocalMailboxListeners)
+            .filter(mailboxListener -> mailboxListener.getExecutionMode().equals(MailboxListener.ExecutionMode.SYNCHRONOUS))
+            .flatMap(mailboxListener -> Mono.fromRunnable(Throwing.runnable(() -> mailboxListener.event(event)))
+                .doOnError(e -> LOGGER.error("Exception happens when handling event of user {}", event.getUser().asString(), e))
+                .onErrorResume(e -> Mono.empty()))
+            .then().cache();
```
Comment (chibenwa) [inline mailbox/event/event-rabbitmq/src/main/java/org/apache/james/mailbox/events/EventDispatcher.java]: Is this `.cache()` required?

File: mailbox/event/event-rabbitmq/src/main/java/org/apache/james/mailbox/events/EventDispatcher.java [hunk 7]
```diff
@@ -80,4 +105,11 @@ void start() {
         return eventSerializer.toJson(event).getBytes(StandardCharsets.UTF_8);
     }
 
+    private OutboundMessage generateMessage(RoutingKeyConverter.RoutingKey routingKey, byte[] payload) {
+        AMQP.BasicProperties properties = new AMQP.BasicProperties.Builder()
```
Comment (chibenwa) [inline mailbox/event/event-rabbitmq/src/main/java/org/apache/james/mailbox/events/EventDispatcher.java]: These AMQP properties are the same for each messages hence maybe it makes little sense to always recompute it. 

Couldn't this be a class field?

File: mailbox/event/event-rabbitmq/src/main/java/org/apache/james/mailbox/events/KeyRegistrationHandler.java [hunk 8]
```diff
@@ -107,18 +109,29 @@ Registration register(MailboxListener listener, RegistrationKey key) {
...
         Event event = toEvent(delivery);
 
         return mailboxListenerRegistry.getLocalMailboxListeners(registrationKey)
+            .filter(listener -> !isLocalSynchronousListeners(eventBusId, listener))
             .flatMap(listener -> Mono.fromRunnable(Throwing.runnable(() -> listener.event(event)))
                 .doOnError(e -> LOGGER.error("Exception happens when handling event of user {}", event.getUser().asString(), e))
-                .onErrorResume(e -> Mono.empty()))
-            .subscribeOn(Schedulers.elastic())
+                .onErrorResume(e -> Mono.empty())
+                .then().cache())
```
Comment (chibenwa) [inline mailbox/event/event-rabbitmq/src/main/java/org/apache/james/mailbox/events/KeyRegistrationHandler.java]: Why do we need this `.then().cache()` ?

File: mailbox/event/event-rabbitmq/src/main/java/org/apache/james/mailbox/events/KeyRegistrationHandler.java [hunk 9]
```diff
@@ -107,18 +109,29 @@ Registration register(MailboxListener listener, RegistrationKey key) {
...
+            .subscribeOn(Schedulers.parallel())
```
Comment (chibenwa) [inline mailbox/event/event-rabbitmq/src/main/java/org/apache/james/mailbox/events/KeyRegistrationHandler.java]: Why did we change to the `parallel` one?

Don't we expect the listeners to be doing some crappy blocking stuff that would justify the elastic scheduler?

File: mailbox/event/event-rabbitmq/src/test/java/org/apache/james/mailbox/events/EventBusIdTest.java [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+    @Test
+    void eventBusIdShouldMatchBeanContract() {
+        EqualsVerifier.forClass(EventBusId.class);
+    }
+
+    @Test
+    void ofShouldDeserializeUUIDs() {
+        assertThat(EventBusId.of(UUID_1.toString()))
+            .isEqualTo(EventBusId.of(UUID_1));
+    }
```
Comment (chibenwa) [inline mailbox/event/event-rabbitmq/src/test/java/org/apache/james/mailbox/events/EventBusIdTest.java]: Can I have a little test for `asString` ?

File: mailbox/event/event-rabbitmq/src/main/java/org/apache/james/mailbox/events/KeyRegistrationHandler.java (hunk 8 above)
Comment (hoangdat) [reply mailbox/event/event-rabbitmq/src/main/java/org/apache/james/mailbox/events/KeyRegistrationHandler.java]: need to add `then()` follow by [discussion](https://chat.linagora.com/openpaas/pl/1y4n7684ety77cz9zungktqgzo)

**Pull request closed without merging**